accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+declarations ever calls `foo`.
+
+Another example would be that we're using interop with JavaScript, and
+we wish to work on a given `JSObject` representing a button, using a
+`Button` interface which is meaningful for buttons. In this case the
+implementation of the members of `Button` will call some low-level
+functions like `js_util.getProperty`, but a client who uses the inline
+class will have a full implementation of the `Button` interface, and
+will hence never need to call `js_util.getProperty`.
+
+(We _can_ just call `js_util.getProperty` anyway, because it accepts
+two arguments of type `Object`. But we assume that the developer will
+be happy about sticking to the rule that the low-level functions
+aren't invoked in application code, and they can do that by using inline
+classes like `Button`. It is then easy to `grep` your application code
+and verify that it never calls `js_util.getProperty`.)
+
+Another potential application would be to generate inline class
+declarations handling the navigation of dynamic object trees that are
+known to satisfy some sort of schema outside the Dart type system. For
+instance, they could be JSON values, modeled using `num`, `bool`,
+`String`, `List<dynamic>`, and `Map<String, dynamic>`, and those JSON
+values might again be structured according to some schema.
+
+Without inline types, the JSON value would most likely be handled with
+static type `dynamic`, and all operations on it would be unsafe. If the
+JSON value is assumed to satisfy a specific schema, then it would be
+possible to reason about this dynamic code and navigate the tree correctly
+according to the schema. However, the code where this kind of careful
+reasoning is required may be fragmented into many different locations, and
+there is no help detecting that some of those locations are treating the
+tree incorrectly according to the schema.
+
+If inline classes are available then we can declare a set of inline types
+with operations that are tailored to work correctly with the given
+schema and its subschemas. This is less error-prone and more
+maintainable than the approach where the tree is handled with static
+type `dynamic` everywhere.
+
+Here's an example that shows the core of that scenario. The schema that
+we're assuming allows for nested `List<dynamic>` with numbers at the
+leaves, and nothing else.
+
+```dart
+inline class TinyJson {
+  final Object it;
+  TinyJson(this.it);
+  Iterable<num> get leaves sync* {
+    if (it is num) {
+      yield it;
+    } else if (it is List<dynamic>) {
+      for (var element in it) {
+        yield* TinyJson(element).leaves;
+      }
+    } else {
+      throw "Unexpected object encountered in TinyJson value";
+    }
+  }
+}
+
+void main() {
+  var tiny = TinyJson(<dynamic>[<dynamic>[1, 2], 3, <dynamic>[]]);
+  print(tiny.leaves);
+  tiny.add("Hello!"); // Error.
+}
+```
+
+Note that `it` is subject to promotion in the above example. This is safe
```

Comment:
I'm ok with this as written.  I suppose you could make an argument that for consistency with the rest of the field promotion work, we shouldn't allow promotion of the representation "field" if it has a public name.  But I'd prefer to go ahead and make representation fields always promotable, because there's no soundness problem AFAICT, and it's easy to do implementation-wise.

Reply:
We could rely on the fact that inline member resolution is static. The reason why promotion cannot soundly be applied to getter invocations that resolve statically to a final instance variable is that it could be overridden by an unstable getter. Not so here. We could immediately do the same thing for a final instance variable in a final class.

Reply:
No objections were raised here. Resolving the thread.